src/2025h2/autoreborrow-traits.md

```diff
@@ -0,0 +1,139 @@
+
+### The "shiny future" we are working towards
+
+`Pin` ergonomics group should be able to get rid of special-casing of `Pin` reborrowing in rustc.
+
+Rust for Linux project should be enabled to experiment with custom reborrowable reference types in earnest.
+
+Users of `reborrow` crate and similar should be enabled to move to core solution.
+
+## Design axioms
+
+- Accept the definition of reborrowing as "a memory copy with added lifetime analysis".
+  - This disallows running user code on reborrow.
+  - "Reborrow-as-shared" likely needs to run user code; this'd preferably be ignored where possible.
+- Must achieve true reborrowing, not a fascimile.
+  - Current userland reborrowing uses `T::reborrow_mut` functions that achieve the most important part of
+    reborrowing, temporarily disabled `T` upon reborrow.
+  - Userland cannot achieve true reborrowing: true reborrowing does not constrain the lifetime of `T`,
+    whereas userland fascimile does.
+  - Difference is in whether values derived from a reborrow can be returned past the point of the reborrow.
+- Performance of the solution must be trivial: reborrow is checked for at every coercion site. This cannot be
+  slow.
+- Make sure autoreborrowing doesn't become a vehicle for implicit type coercion. Allowing autoreborrowing
+  from `T` to multiple values could be abused to define a `CustomInt(int128)` that coerces to all integer
+  types.
+  - Autoreborrow traits should use an associated type instead of a type parameter.
+  - Autoreborrowing at coercion sites should not dovetail into eg. an `Into::into` call.
+
+## Ownership and team asks
+
+| Task                         | Owner(s) or team(s) | Notes |
+|------------------------------|---------------------|-------|
+| Discussion and moral support | ![Team][] [lang]    | Normal RFC process |
+| Help around impl             | ![Team][] [compiler]| Trait-impl querying in rustc to replace `Pin<&mut T>` special case |
+| Do the work                  | @aapoalas           |       |
+
+### Design autoreborrow internals
+
+The basic idea of autoreborrowing is simple enough: when a reborrowable type is encountered at a coercion
+site, attempt a reborrow operation.
+
+Complications arise when reborrowing becomes recursive: if a `struct X { a: A, b: B }` contains two
+reborrowable types `A` and `B`, then we'd want the reborrow of `X` to be performed "piecewise". As an
+example, the following type should, upon reborrow, only invalidate any values that depend on the `'a` lifetime while any values dependent on the `'b` lifetime should still be usable as normal.
+
+```rust
+struct X<'a, 'b> {
+    a: &'a mut A,
+    b: &'b B,
+}
+```
+
+To enable this, reborrowing needs to be defined as a recursive operation but what the "bottom-case" is, that
+is the question. One option would be to use `!Copy + Reborrow` fields, another would use core marker types
+like `PhantomExclusive<'a>` and `PhantomShared<'b>` to discern the difference.
```

Comment:
Would an attribute on the fields themselves work?

I would guess most structs would only have one field that gets reborrowed at a time, but some would have more than one. Does that sound right to you?